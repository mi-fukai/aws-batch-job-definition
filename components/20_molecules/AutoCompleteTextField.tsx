import React, { useState } from "react";
import { TextField } from "@material-ui/core";
import { Autocomplete } from "@material-ui/lab";

import { useDelayedEffect } from "@/lib/customHooks";

interface TypeProps {
  searchText: string;
  setSearchText: React.Dispatch<React.SetStateAction<string>>;
}

export const AutoCompleteTextField: React.VFC<TypeProps> = ({
  searchText,
  setSearchText,
}: TypeProps) => {
  const defaultAutoCompleteKeyWords: string[] = [
    "hoge",
    "hogehoge",
    "hogehogehoge",
    "hogehogehogehoge",
  ];

  const [autoCompleteKeyWords, setAutoCompleteKeyWords] = useState(
    defaultAutoCompleteKeyWords
  );
  const updateSearchText = (event, newValue: string) => {
    setSearchText(newValue);
  };
  const updateAutoCompleteKeyWords = async (): Promise<void> => {
    // {result:[{text:"AutoCompleteKeyWord"}],...}
    if (searchText) {
      const autoCompleteKeyWords: string[] = (
        await (
          await fetch(
            `${process.env.NEXT_PUBLIC_HOST}/api/suggest/${searchText}`
          )
        ).json()
      ).result.map((item) => item.text);
      setAutoCompleteKeyWords(autoCompleteKeyWords);
    } else {
      setAutoCompleteKeyWords(defaultAutoCompleteKeyWords);
    }
  };

  useDelayedEffect(() => {
    updateAutoCompleteKeyWords();
  }, [searchText]);

  return (
    <Autocomplete
      fullWidth
      value={searchText}
      onInputChange={updateSearchText}
      options={autoCompleteKeyWords}
      freeSolo
      renderInput={(params) => (
        <TextField {...params} label="きーわーど" />
      )}
    ></Autocomplete>
  );
};

export default AutoCompleteTextField;
