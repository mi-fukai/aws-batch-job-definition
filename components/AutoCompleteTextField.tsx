import React, { useState } from "react";
import { TextField } from "@material-ui/core";
import { Autocomplete } from "@material-ui/lab";

import { useDelayedEffect } from "@/lib/customHooks";

interface TypeProps {
  props: {
    searchText: string;
    setSearchText: React.Dispatch<React.SetStateAction<string>>;
  };
}

export const AutoCompleteTextField: React.VFC<TypeProps> = ({
  props,
}: TypeProps) => {
  const { searchText, setSearchText } = props;
  const defaultAutoCompleteKeyWords: string[] = [
    "hoge",
    "hogehoge",
    "hogehogehoge",
    "hogehogehogehoge",
  ];

  const [autoCompleteKeyWords, setAutoCompleteKeyWords] = useState(
    defaultAutoCompleteKeyWords
  );
  const updateSearchText = (event, newValue: string) => {
    setSearchText(newValue);
  };
  const updateAutoCompleteKeyWords = async (): Promise<void> => {
    // {result:[{text:"AutoCompleteKeyWord"}],...}
    if (searchText) {
      const autoCompleteKeyWords: string[] = (
        await (
          await fetch(
            `${process.env.NEXT_PUBLIC_HOST}/api/suggest/${searchText}`
          )
        ).json()
      ).result.map((item) => item.text);
      setAutoCompleteKeyWords(autoCompleteKeyWords);
    } else {
      setAutoCompleteKeyWords(defaultAutoCompleteKeyWords);
    }
  };

  useDelayedEffect(() => {
    updateAutoCompleteKeyWords();
  }, [searchText]);

  return (
    <Autocomplete
      value={searchText}
      onInputChange={updateSearchText}
      options={autoCompleteKeyWords}
      freeSolo
      renderInput={(params) => (
        <TextField {...params} id="standard-basic" label="きーわーど" />
      )}
    ></Autocomplete>
  );
};
