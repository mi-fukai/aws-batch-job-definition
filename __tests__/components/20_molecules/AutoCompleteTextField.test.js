import React, { useState } from "react";
import { cleanup, render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";
import AutoCompleteTextField from "@/components/20_molecules/AutoCompleteTextField";


afterEach(cleanup);

describe("default items", () => {
  test("searchText props is textField default value", () => {
    const randomText = Math.random().toString(32).substring(2);
    const renderUseStateHook = renderHook(
      (props) => {
        const [searchText, setSearchText] = useState(props);
        return [searchText, setSearchText];
      },
      { initialProps: randomText }
    );
    const [searchText, setSearchText] = renderUseStateHook.result.current;
    const { getByRole } = render(
      <AutoCompleteTextField
        searchText={searchText}
        setSearchText={setSearchText}
      />,
      {}
    );
    expect(getByRole("textbox").value).toBe(randomText);
  });
});

describe("Snap Shot test", () => {
  it("Component Global Snap Shot", () => {
    // TextFieldのidランダム生成をされるとスナップショットが失敗するのでそれを回避する
    const mockMath = Object.create(global.Math);
    mockMath.random = () => 0.5;
    global.Math = mockMath;

    const { asFragment } = render(<AutoCompleteTextField />, {});
    expect(asFragment()).toMatchSnapshot();
  });
});
