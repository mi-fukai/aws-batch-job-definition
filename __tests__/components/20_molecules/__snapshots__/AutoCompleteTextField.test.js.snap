// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snap Shot test Component Global Snap Shot 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="mui-50000"
        id="mui-50000-label"
      >
        きーわーど
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
          id="mui-50000"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
